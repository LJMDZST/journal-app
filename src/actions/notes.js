
import Swal from 'sweetalert2';
import {db} from '../firebase/firebase_config.js'
import { fileUpload } from '../helpers/fileUpload.js';
import { loadNotes } from '../helpers/loadNotes.js';
import {types} from '../types/types.js'

// Cargar un nuevo registro en la Base de Datos..

export const startNewNote = ()=>{
    return async(dispatch, getState) =>{
        const {uid} = getState().auth;
        
        const newNote = {
            title: '',
            body : '',
            date : new Date().getTime(),
        }
        try {
            const doc = await db.collection(`${uid}/journal/notes`).add( newNote )

            dispatch( activeNote(doc.id, newNote ) );
            dispatch( addNewNote(doc.id, newNote ) );
            
        } catch (error) {
            console.log( error );
        }
        
    }
}

export const activeNote = (id,note)=>({
    type : types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const addNewNote = (id,note)=>({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes =(uid)=> {

    return async(dispatch)=>{

        const notes = await loadNotes( uid );

        dispatch( setNotes( notes ) );
    }
}

export const setNotes = ( notes )=>({
    type: types.notesLoad,
    payload: notes
})


export const startSaveNote = (note)=>{
    return async(dispatch,getState)=>{
        const {uid} = getState().auth;

        if( !note.url){
            delete note.url;
        }


        const noteToFirestore = {...note}; // copia del objeto
        delete noteToFirestore.id;

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
    
        dispatch( refreshNote(note.id, note) );

        Swal.fire('Saved :)', note.title, 'success');

    }
}


export const refreshNote = (id,note)=>({
    type: types.notesUpdated,
    payload: {
        id,
        note
    }
})


export const startUpload = (file)=>{
    return async(dispatch, getState)=>{
        const { active:activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Por favor , espere :)',
            allowOutsideClick: false,
            onBeforeOpen: ()=>{
                Swal.showLoading();
            }
        });
        try {
            const fileUrl = await fileUpload(file);
            activeNote.url = fileUrl;
            dispatch( startSaveNote( activeNote));
    
            Swal.close();
        } catch (error) {
            console.log(error)
        }
       
    }
}

export const startDeleting = ( id )=>{
    return async(dispatch, getState)=>{
        const uid = getState().auth.uid;
        await db.doc(`/${ uid }/journal/notes/${ id }`).delete()
            .then(()=> Swal.fire('Suces','Deleted :)','success'))
            .catch(e => Swal.fire('Error',e,'error'));
        dispatch( deleteNote(id));
    }
}

export const deleteNote = (id)=>({
    type: types.notesDelete,
    payload : id
})

export const noteLogout = ()=>({
    type: types.notesLogoutCleaning
    
})