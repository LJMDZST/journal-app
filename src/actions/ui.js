import { types } from "../types/types"



export const mostrarError = (err) => ({
    type : types.uiSetError,
    payload : err
})

export const limpiarError = () => ({
    type : types.uiRemoveError
})

export const uiStartLoading = () => ({
    type : types.uiStartLoading
})

export const uiFinishLoading = () => ({
    type : types.uiFinishLoading
})