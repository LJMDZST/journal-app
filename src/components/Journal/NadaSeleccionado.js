import React from 'react'

/** Componente de inicio, que avisa que no hay notas en la aplicacion. */

export const NadaSeleccionado = () => {
    return (
        <div className="nothing__main-control animate__animated animate__fadeInRight">
            <p>
                Select something
                <br />
                pr create an entry!
            </p>
            <i className="far fa-star fa-4x mt-5"></i>
        </div>
    )
}
