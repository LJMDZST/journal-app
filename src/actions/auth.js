import Swal from 'sweetalert2';
import {firebase, googleAuthProvider} from '../firebase/firebase_config'
import { types } from "../types/types"
import { noteLogout } from './notes';
import { uiFinishLoading, uiStartLoading } from './ui'

/** Funciones asincronas para las acciones de autenticacion */


// Accion de boton de logueo con email y password
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) =>{
        dispatch( uiStartLoading()  )
        return firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user}) => {
                dispatch( login(user.uid,user.displayName) )
                dispatch( uiFinishLoading() )
            })
            .catch((e)=>{
                dispatch( uiFinishLoading() );
                Swal.fire('Error !',e.message,'error');
            });
    }
}

// accion para registrar un nuevo ususario con contraseÃ±a
export const startRegisterWithEmailPasswordName = (email,password,name) =>{

    return (dispatch) =>{
        return firebase.auth().createUserWithEmailAndPassword(email,password)
                .then( async({ user }) =>{
                    // para actualizar el displayNAme o la foto de usuario
                    // en firebase
                    await user.updateProfile({displayName : name});
                    // console.log(user)

                     dispatch (
                        
                        login( user.uid, user.displayName )
                     )
                } )
                .catch((e)=>{
                    console.log(e);
                    Swal.fire('Error !',e.message,'error');
                });
    }


}

// Accion de boton de logueo con cuenta de google
export const startGoogleLogin = () =>{
    return ( dispatch )=>{
       return firebase.auth().signInWithPopup( googleAuthProvider )
                .then(  ({ user }) =>{
                    dispatch (
                        login( user.uid, user.displayName )
                    )
                } );
    }
}

/** accion de logueo de usuario
 *  -> retorna un objeto con los datos del usuario logueado
 */
export const login = (uid, displayName) => ({
    type : types.login,
    payload : {
        uid,
        displayName
    }
});


export const startLogout = () =>{
    return async(dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout());
        dispatch(noteLogout());
    }
}
    
       
export const  logout = () =>({
    type : types.logout
})    
