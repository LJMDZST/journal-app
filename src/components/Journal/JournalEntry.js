import moment from 'moment'
import React from 'react'
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';


/** Componente Item del listado de Notas del menu lateral, */
export const JournalEntry = ({id,date,title,body,url}) => {
    
    const noteDate = moment( date );
    const dispatch = useDispatch();

    const handleEntryClick = ()=>{
        dispatch( activeNote( id, {
            title,
            body,
            date,
            url
        } ) )
    }

    return (
        <div className="journal__entry animate__animated animate__fadeInDown"
             onClick={handleEntryClick}
        >
            {  url &&
                <div className="journal__entry-picture"
                    style={{
                        backgroundSize: 'cover',
                        backgroundImage: `url(${url})`,
                    }}
                ></div>
            }

            <div className="journal__entry-body">
                <p className="journal__entry-title">
                   {title}
                </p>
                <p className="journal__entry-content">
                    {body}
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span>{noteDate.format('dddd')}</span>
                <h4>{noteDate.format('Do')}</h4>
            </div>
        </div>
    )
}
