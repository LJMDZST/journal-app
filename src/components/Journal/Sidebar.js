import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries'


/** MenÃº lateral de la app, 
 *  Tiene el nombre de usario y el boton de cerrar sesion.
 *  Tiene un boton para agregar una nueva nota.
 *  Tiene un listado del historial de notas.
 */
export const Sidebar = () => {
    
    const state = useSelector(state => state.auth)
    const dispatch = useDispatch();

    const handleAddNew = ()=>{
        dispatch( startNewNote() );
    }

    const handleLogout = ()=>{
        dispatch(startLogout());
    }


    return (
        <aside className="journal__sidebar">


            <div className="journal__sidebar-navbar">
                <h3 className="mt-5">
                    <i className="far fa-moon"></i>
                    <span> {state.name}</span>
                </h3>

                <button className="btn"
                    onClick={handleLogout}
                >
                    Logout
                </button>
            </div>

            <div className="journal__new-entry"
                 onClick={handleAddNew}
            >
                <i className="far fa-calendar-plus fa-5x"></i>
                <p className="mt-5">
                    New Entry
                </p>
            </div>
            
            <JournalEntries />

        </aside>
    )
}
