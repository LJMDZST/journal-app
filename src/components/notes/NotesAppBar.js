import moment from 'moment'
import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUpload } from '../../actions/notes';


/**Barra de Cabecera del espacio de nota Tiene una fecha y dos botones */
export const NotesAppBar = () => {

    
    const {active} = useSelector(state => state.notes)

    const noteDate = moment(active.date);
    const dispatch = useDispatch();



    const handlePictureUpload = ()=>{
        document.querySelector('#fileSelector').click();
    }

    const handleChangeFile =(e)=>{
        const file = e.target.files[0];

        dispatch ( startUpload( file ) );
    }
    
    const handleSave = ()=>{
        dispatch( startSaveNote(active))
    }

    
    
    return (
        <div className="notes_appbar">
            <span> {noteDate.format("dddd Do, MMMM YYYY")} </span>

            <input 
                id="fileSelector"
                name="file"
                type="file"
                style={{display:'none'}}
                onChange={handleChangeFile}
                
            />

            <div>
                <button className="btn"
                    onClick={handlePictureUpload}
                >
                    Picture
                </button>
                <button className="btn"
                    onClick={handleSave}
                >
                    Save
                </button>
            </div>
        </div>
    )
}
