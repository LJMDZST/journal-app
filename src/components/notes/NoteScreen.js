import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'


/** Seccion de nota :
 *  Tiene una cabecera (fecha, seleccionar foto y guardar) <NotesAppBar/>
 *  Tiene un espacio de escritura( Titulo, Contenido) 
 */
export const NoteScreen = () => {

    const {active:note} = useSelector(state => state.notes);
    const dispatch = useDispatch();

    const [formValues, handleInputchange,reset] = useForm(note);

    const {body,title,url} = formValues;

    const activeId = useRef(note.id);

    useEffect(() => {
        
        if(note.id !== activeId.current){
            reset(note);
            activeId.current = note.id;
        }


    }, [note,reset])


    useEffect(() => {
        
        dispatch( activeNote(formValues.id,{...formValues}) );

    }, [formValues,dispatch])

    const handleDelete = ()=>{

        dispatch( startDeleting( note.id ) );
    }

    return (
        <div className="notes__main-content animate__animated animate__fadeIn">
            <NotesAppBar />

            <div className="notes__content" >
                <input className="notes__title-input" 
                    name="title"
                    type="text"
                    placeholder="Some awesome title"
                    value={title}
                    onChange={handleInputchange}
                />

                <textarea className="notes__textarea"
                    name="body"
                    placeholder="Whats happened today :)??"
                    value={body}
                    onChange={handleInputchange}
                ></textarea>


                {   note.url &&
                    <div className="notes__image">
                        <img 
                            src={url}
                            alt="imagen"
                        />
                    </div>

                }

                <button className="btn btn-danger"
                    onClick = {handleDelete}
                >Delete</button>
                
            </div>
        </div>
    )
}
