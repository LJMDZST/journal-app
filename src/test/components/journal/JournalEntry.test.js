import React from 'react';
import ReactDOM from 'react-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import '@testing-library/jest-dom';


import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { JournalEntry } from '../../../components/Journal/JournalEntry';
import { activeNote } from '../../../actions/notes';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
    id: 0,
    title: 'Hola',
    date: 0,
    body: 'Mundo',
    url: 'http://imagen.com/imagen.jpg'
}

const wrapper = mount(
    <Provider store={store}>
        <JournalEntry {...note} />
    </Provider>
);



describe('Pruebas en JorunalEntry', () => {
    
    test(' debe mostrar correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();


    });

    test('debe activar la nota', () => {
        
        wrapper.find( '.journal__entry').prop( 'onClick')();

        expect( store.dispatch ).toHaveBeenCalledWith(
            activeNote( note.id , { ...note } )
        );

    })
    
    

})
