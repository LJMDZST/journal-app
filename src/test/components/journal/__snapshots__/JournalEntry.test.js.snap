// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en JorunalEntry  debe mostrar correctamente 1`] = `
<div
  className="journal__entry animate__animated animate__fadeInDown"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(http://imagen.com/imagen.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hola
    </p>
    <p
      className="journal__entry-content"
    >
      Mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
